### PR TITLE
feature: Add duplication tests

### PR DESCRIPTION
**Pull request by lolgab**

### CODE REVIEW

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  4
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  3
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
         

Clones added
============
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  1
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
- src/main/scala/codacy/plugins/test/multiple/ResultPrinter.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

File: src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package codacy.plugins.test.checkstyle
+
+import scala.xml.NodeSeq
+
+object CheckstyleImplicits {
+  implicit class NodeSeqOps(val node: NodeSeq) extends AnyVal {
+
+    def getProperty(name: String): String = {
+      (node \ "property").find(_ \@ "name" == name).get \@ "value"
```
Comment (mrfyda) [inline src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala]: Codacy Issue found: [Usage of get on optional type.](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117)

File: src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        debug(s"${testDirectory.name} should have ${expectedResults.size} results")
+        duplicationTools.exists { tool =>
+          val res = runTool(tool, srcDir)
+          ResultPrinter.printToolResults(res, expectedResults)
+        }
+      }
+      .forall(identity)
+  }
+
+  private def runTool(tool: core.tools.DuplicationTool, multipleTestsDirectory: File): Try[Set[DuplicationClone]] = {
```
Comment (josemiguelmelo) [inline src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala]: `multipleTestsDirectory` should be called `duplicationTestsDirectory`

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  3
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  4
         

Clones added
============
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  1
- src/main/scala/codacy/plugins/test/multiple/ResultPrinter.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

File: src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala (hunk 1 above)
Comment (mrfyda) [inline src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala]: Codacy Issue found: [Usage of get on optional type.](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117)

File: src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        } else {
+          error("Tool results don't match expected results:")
+          error("Extra: ")
+          info(pprint.apply(results.diff(expectedResults), height = Int.MaxValue))
+          error("Missing:")
+          info(pprint.apply(expectedResults.diff(results), height = Int.MaxValue))
+          false
+        }
+    }
+}
```
Comment (josemiguelmelo) [inline src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala]: Isn't this `ResultPrinter` duplicated? Why not using an abstract type on the set?

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  3
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  4
         

Clones added
============
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
- src/main/scala/codacy/plugins/test/duplication/ResultPrinter.scala  1
- src/main/scala/codacy/plugins/test/multiple/ResultPrinter.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

File: src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala (hunk 1 above)
Comment (mrfyda) [inline src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala]: Codacy Issue found: [Usage of get on optional type.](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117)

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala  2
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  3
- src/main/scala/codacy/plugins/test/resultprinter/ResultPrinter.scala  4
         

Clones added
============
- src/main/scala/codacy/plugins/test/multiple/MultipleTests.scala  1
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  3
- src/main/scala/codacy/plugins/test/resultprinter/ResultPrinter.scala  4
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala  2
         

Clones added
============
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
- src/main/scala/codacy/plugins/test/multiple/MultipleTests.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/resultprinter/ResultPrinter.scala  4
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  2
- src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala  2
         

Clones added
============
- src/main/scala/codacy/plugins/test/multiple/MultipleTests.scala  1
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala  2
- src/main/scala/codacy/plugins/test/resultprinter/ResultPrinter.scala  4
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  2
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
         

Clones added
============
- src/main/scala/codacy/plugins/test/multiple/MultipleTests.scala  1
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

**Comment by mrfyda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/scala/codacy/plugins/test/duplication/CheckstyleFormatParser.scala  2
- src/main/scala/codacy/plugins/test/resultprinter/ResultPrinter.scala  4
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  2
- src/main/scala/codacy/plugins/test/checkstyle/CheckstyleImplicits.scala  2
         

Clones added
============
- src/main/scala/codacy/plugins/test/duplication/DuplicationTests.scala  1
- src/main/scala/codacy/plugins/test/multiple/MultipleTests.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/codacy/codacy-plugins-test/pullRequest?prid=4860117&bid=16123031)

**Comment by lolgab** [comment]
Good job @josemiguelmelo :clap:

**Pull request merged**